general/arrays_by_name.hpp

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2010-2023, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_ARRAYS_BY_NAME
+#define MFEM_ARRAYS_BY_NAME
+
+#include "../config/config.hpp"
+#include "array.hpp"
+
+#include <iostream>
+#include <map>
+#include <set>
+#include <string>
+
+namespace mfem
+{
+
+template <class T>
+class ArraysByName
+{
+protected:
+   /// Reusing STL map iterators
+   using container      = std::map<std::string,Array<T> >;
+   using iterator       = typename container::iterator;
+   using const_iterator = typename container::const_iterator;
+
+   /// Map containing the data sorted alphabetically by name
+   container data;
+
+public:
+
+   /// @brief Default constructor
+   ArraysByName() = default;
+
+   /// @brief Copy constructor: deep copy from @a src
+   ArraysByName(const ArraysByName &src) = default;
+
+   /// @brief Move constructor: deep move from @a src
+   ArraysByName(ArraysByName &&src) noexcept = default;
+
+   /// @brief Copy constructor (deep copy) from 'src', an ArraysByName
+   /// container of convertible type.
+   template <typename CT>
+   inline ArraysByName(const ArraysByName<CT> &src);
```

Comment:
I know this is analogous to a similar constructor in `Array<T>`, but I really don't think this is needed ([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)). If down the line we have a need for something like this, we can always add it. (The downside of adding something now rather than waiting until later is that we have to commit to backwards compatibility, maintaining this code in the future, etc.).